haskell/ghc_bindist_hadrian.bzl

```diff
@@ -0,0 +1,439 @@
+    target_compatible_with = {target_constraints},
+)
+        """.format(
+            bindist_name = ctx.attr.bindist_name,
+            exec_constraints = exec_constraints,
+            target_constraints = target_constraints,
+        ),
+    )
+
+_ghc_bindist_hadrian_toolchain = repository_rule(
+    _ghc_bindist_hadrian_toolchain_impl,
+    local = False,
+    attrs = {
+        "bindist_name": attr.string(),
+        "target": attr.string(),
+    },
+)
+
+def ghc_bindist_hadrian(
+        name,
+        url,
+        sha256,
+        strip_prefix,
+        version,
+        target,
+        ghcopts = None,
+        haddock_flags = None,
+        repl_ghci_args = None,
+        cabalopts = None,
+        locale = None):
+
+    bindist_name = name
+    toolchain_name = "{}-toolchain".format(name)
+
+    # We want the toolchain definition to be tucked away in a separate
+    # repository, that way `bazel build //...` will not match it (and
+    # e.g. build the Windows toolchain even on Linux). At the same
+    # time, we don't want the definition in the bindist repository,
+    # because then we need to download the bindist first before we can
+    # see the toolchain definition. The solution is to add the
+    # toolchain definition in its own special repository.
+    _ghc_bindist_hadrian(
+        name = bindist_name,
+        url = url,
+        sha256 = sha256,
+        strip_prefix = strip_prefix,
+        version = version,
+        ghcopts = ghcopts,
+        haddock_flags = haddock_flags,
+        repl_ghci_args = repl_ghci_args,
+        cabalopts = cabalopts,
+        target = target,
+        locale = locale,
+    )
+    _ghc_bindist_hadrian_toolchain(
+        name = toolchain_name,
+        bindist_name = bindist_name,
+        target = target,
+    )
+    native.register_toolchains("@{}//:toolchain".format(toolchain_name))
+
+def haskell_register_ghc_bindists_hadrian(
+        url,
+        sha256,
+        strip_prefix,
+        version,
+        target,
+        ghcopts = None,
+        haddock_flags = None,
+        repl_ghci_args = None,
+        cabalopts = None,
+        locale = None):
+    ghc_bindist_hadrian(
+        name = "rules_haskell_ghc_{}".format(target),
+        url = url,
+        sha256 = sha256,
+        strip_prefix = strip_prefix,
+        target = target,
+        version = version,
+        ghcopts = ghcopts,
+        haddock_flags = haddock_flags,
+        repl_ghci_args = repl_ghci_args,
+        cabalopts = cabalopts,
+        locale = locale,
+    )
+    local_sh_posix_repo_name = "rules_haskell_sh_posix_local"
+    if local_sh_posix_repo_name not in native.existing_rules():
+        sh_posix_configure(name = local_sh_posix_repo_name)
+    local_python_repo_name = "rules_haskell_python_local"
+    if local_python_repo_name not in native.existing_rules():
+        _configure_python3_toolchain(name = local_python_repo_name)
+
+def _configure_python3_toolchain_impl(repository_ctx):
```

Comment:
Fine for development, but we'll want to clean this up and deduplicate with `haskell/ghc_bindist.bzl`.

Reply:
Right.